doc/internals.rst

```diff
@@ -231,3 +231,200 @@ re-open it directly with Zarr:
+
+If it makes sense for your backend, your ``open_dataset`` method should
+implement in its interface all the following boolean keyword arguments, called
+**decoders** which default to ``None``:
+
+- ``mask_and_scale=None``
+- ``decode_times=None``
+- ``decode_timedelta=None``
+- ``use_cftime=None``
+- ``concat_characters=None``
+- ``decode_coords=None``
+
+These keyword arguments are explicitly defined in Xarray
+:py:meth:`~xarray.open_dataset` signature.  Xarray will pass them to the
+backend only if the User sets a value different from ``None`` explicitly.
+Your backend can also take as input a set of backend-specific keyword
+arguments. All these keyword arguments can be passed to
+:py:meth:`~xarray.open_dataset` grouped either via the ``backend_kwarg``
+parameter or explicitly using the syntax ``**kwargs``.
+
+Output
+^^^^^^
+The output of the backend `open_dataset` shall be an instance of
+Xarray py:class:`~xarray.Dataset` that implements the additional method ``close``,
+used by Xarray to ensure the related files are eventually closed.
+
+If you don't want to support the lazy loading, then the :py:class:`~xarray.Dataset`
+shall contain ``NumPy.arrays`` and your work is almost done.
+
+open_dataset_parameters
++++++++++++++++++++++++
+``open_dataset_parameters`` is the list of backend ``open_dataset`` parameters.
+It is not a mandatory parameter, and if the backend does not provide it
+explicitly, Xarray creates a list of them automatically by inspecting the
+backend signature.
+
+Xarray uses ``open_dataset_parameters`` only when it needs to select
+the **decoders** supported by the backend.
+
+If ``open_dataset_parameters`` is not defined, but ``**kwargs`` and ``*args`` have
+been passed to the signature, Xarray raises an error.
+On the other hand, if the backend provides the ``open_dataset_parameters``,
+then ``**kwargs`` and `*args`` can be used in the signature.
+
+However, this practice is discouraged unless there is a good reasons for using
+`**kwargs` or `*args`.
+
+guess_can_open
+++++++++++++++
+``guess_can_open`` is used to identify the proper engine to open your data
+file automatically in case the engine is not specified explicitly. If you are
+not interested in supporting this feature, you can skip this step since
+py:class:`~xarray.backend.common.BackendEntrypoint` already provides a default
+py:meth:`~xarray.backend.common BackendEntrypoint.guess_engine` that always returns ``False``.
+
+Backend ``guess_can_open`` takes as input the ``filename_or_obj`` parameter of
+Xarray :py:meth:`~xarray.open_dataset`, and returns a boolean.
+
+
+How to register a backend
++++++++++++++++++++++++++++
+
+Define in your setup.py (or setup.cfg) a new entrypoint with:
+
+- group: ``xarray.backend``
+- name: the name to be passed to :py:meth:`~xarray.open_dataset`  as ``engine``.
+- object reference: the reference of the class that you have implemented.
+
```

Comment:
it would probably help to have a example here, too

Reply:
added